DEVELOPMENT.md

```diff
@@ -310,6 +333,361 @@ Incompatible changes require incrementing the major version number. In our curre
+For example, `ControlledGcpResourceApiControllerBqDatasetTest.java` has many
+tests that clone a BQ dataset. The source BQ dataset is initialized once in
+`setup()`, and reused for many tests.
+
+#### For local runs, skip workspace/context creation
+
+Say I'm running `ControlledGcpResourceApiControllerBqDatasetTest.java` which has
+10 tests. The previous section made it so that one workspace is created, instead
+of 10. This section makes it so that 0 workspaces are created.
+
+These changes make local development faster. They should not be merged.
+
+- [Optional] [Comment out these lines](https://cs.github.com/DataBiosphere/terra-workspace-manager/blob/05dba30e7f597690c46c95a974d31bde532bcbbd/service/src/main/java/bio/terra/workspace/app/StartupInitializer.java?q=startupinitializer#L33-L37)
+  This way, when you restart postgres container, your workspaces won't be clobbered.
+- Comment out [`cleanup()`](https://github.com/DataBiosphere/terra-workspace-manager/blob/0764301d03814ab13e6cce5291e191201bddd205/service/src/test/java/bio/terra/workspace/app/controller/ControlledGcpResourceApiControllerBqDatasetTest.java#L132-L133) in your test,
+  so workspaces aren't deleted at the end of the test.
+- Run a test so that `setup()` creates workspace(s). Get workspace ID from
+  DB:
+  ```
+  PGPASSWORD=dbpwd psql postgresql://127.0.0.1:5432/wsm_db -U dbuser
+  wsm_db=> select workspace_id from cloud_context;
+  
+  # If you need projectId
+  wsm_db=> select workspace_id, context->>'gcpProjectId' from cloud_context;
+  ```
+- In `setup()`, comment out workspace initialization and add:
+  ```
+  workspaceId = UUID.fromString("workspace-id");
+  ```
+
+Pre-requisites:
+
+- Your test must store `workspaceId`, not `workspace`.
+- Your test must be able to run more than once on the same workspace. For example,
+  instead of creating a resource named `resource-name`, call it `TestUtils.appendRandomNumber("resource-name")`.
+
+### Cleaning up workspaces in tests
+
+We have 2 ways of cleaning up resources (WSM workspace, SAM workspace, GCP project):
+
+1. Connected tests use Janitor. Janitor deletes GCP project and not SAM
+   workspace (see
+   [here](https://github.com/DataBiosphere/terra-workspace-manager/pull/755#discussion_r942717257) for details).
+2. Tests call WSM `deleteWorkspace()`. This deletes WSM workspace + SAM workspace +
+   GCP project.
+
+Connected tests that use mock SamService: Tests don't need to call
+`deleteWorkspace()` because there is no SAM workspace to clean up.
+
+Connected tests that use real SamService: Tests should call `deleteWorkspace()`
+to clean up SAM workspaces. Why not just call `deleteWorkspace()` and not use
+janitor? Janitor is useful in case test fails (or `deleteWorkspace()` fails).
+
+Integration tests: Tests should call `deleteWorkspace()` because integration
+tests don't use janitor. Most tests don't need to worry about this because
+`WorkspaceAllocateTestScriptBase.java` deletes the workspace it creates.
```

Comment:
Could you add a sentence or two about the cases where tests do need to worry about this? In particular we usually have issues with `clone` tests, since they'll create a second workspace but not clean it up

Reply:
Done